text/0000-volta-use.md

```diff
@@ -0,0 +1,103 @@
+
+This has led to some confusion for users. We get questions fairly regularly asking how users should select an "active" version of Node, once they have installed them using `volta install`. Additionally, by centering the install action, rather than the selection, `volta install` doesn't help users internalize Volta's mental model of "select a version and Volta will automatically download and make sure it's available."
+
+Selecting a different verb for the action of choosing a tool version should reduce confusion and make it more clear how Volta is intended to work.
+
+# Pedagogy
+
+[pedagogy]: #pedagogy
+
+This will be a fairly large change to our recommended workflows. As `volta use` will work for both pinning _and_ selecting a default version, we will need to update the documentation throughout to take advantage of the new command. We will also likely want to write a blog post for the release which includes this change, explaining the reasoning and the new workflows.
+
+Notably, we will _not_ be removing the existing `volta install` and `volta pin` commands right away. They will continue to work as they already do. In the future, we might consider adding notifications to them to point users to the updated command, but for the initial implementation we don't want to disrupt current users' workflows.
+
+All that said, the new command should simplify things a lot, especially for new users. It will allow us to explain Volta by saying that you pick a version with `volta use`, and don't worry about installing it: Volta will handle making it available. It also should do a better job of living up to the claim on our website by truly having a _single_ command that users can run to get set up.
+
+# Details
+
+[details]: #details
+
+## Command definition
+
+```
+USAGE:
+    volta use [FLAGS] <tool[@version]>...
+
+FLAGS:
+    -g, --global    Select the global version of a tool, which is used everywhere outside of pinned projects
+    -p, --pin       Pin a version of a tool to the local project, creating a Volta entry in package.json
+
+ARGS:
+    <tool[@version]>... Tools to make available, like `node@lts` or `yarn@^1.14`
+```
+
+## Behavior
+
+The behavior of `volta use` will ultimately be the same as the current `volta install` or `volta pin`. However, when run without flags, it will be contextual, based on where the command is being run. So if a user runs `volta use node@18` in a pinned project, the project will be updated to reflect the newly selected version. If they instead run `volta use node@18` _outside_ of a pinned project, then the default Node version will be updated. This ensures that running `volta use node@version` will always update the version that is currently active, meaning the next run of `node` will use that updated version.
```

Comment:
I am mulling on the design here in general and my reaction to it, and I could probably go either way on the overall collapsing of `install` and `pin` into a single `use` verb, but I *deeply* dislike this kind of contextual behavior as a user of this kind of tool. The reason I dislike it is that I want the same command to behave the same way given the same arguments. This feels like a case of “global mutable state”, i.e. the directory I am running in and whether it happens to have had a `package.json` file added to it or not, affecting the behavior of a function call, and I personally prefer to *avoid* my `static mut` values. 😆 It seems particularly iffy to me in that it *also* gets changed by whether there is a `package.json` in a directory anywhere between me and my home directory at the time it happens.

Consider a hypothetical folder layout:

```
/users/chris/code/
  DefinitelyTyped/
    DefinitelyTyped/
      package.json
      types/
        react/
          package.json
          ...
  volta-cli/
    package.json # notice this file
    rfcs/
    volta/
      src/
      some-node-support-script/
      ...
```

I will get completely different behaviors from `volta use node@lts` in each of these directories:

- `/users/chris/code`: sets the default
- `/users/chris/code/DefinitelyTyped`: sets the default
- `/users/chris/code/DefinitelyTyped/DefinitelyTyped`: sets the DT project
- `/users/chris/code/DefinitelyTyped/DefinitelyTyped/types`: sets the DT project
- `/users/chris/code/DefinitelyTyped/DefinitelyTyped/types/react`: sets the React project w/in DT
- `/users/chris/code/volta-cli/volta/`: sets the `volta-cli` default
- `/users/chris/code/volta-cli/volta/some-node-support-script/`: sets the `volta-cli` default

This is all internall consistent, and it is comprehensible, but I don’t find it *obvious*, specifically because it is *so* non-local. Scripts could of course make sure to always do the right thing by explicitly passing `--pin` or `--global`, but I do not like that they *need* to.

Also, I note don’t think we have any other command that works quite that contextually. `volta list` is close, but it just shows you the full hierarchy or your tools. If you are not in a project, you *only* get default/global-installed tools listed, but you will get those listed, with our shadowing rules applied, when you *are* in a package as well.

🤔

Thinking out loud a bit here: What if we more or less “renamed” `pin` to `use` and added `--default` to it? That eliminates all the ambiguity above, always behaves the same way, and still gets us all the *conceptual* simplification of having a single command that does what it says on the tin. It comes with the slight extra typing cost of `volta use --default node@lts` instead of “just working”. However, I think that tradeoff is probably worth it.

I would not fight *especially* hard about this, and definitely would not block on it, but I definitely prefer this design. Thoughts?

Reply:
To start off, I agree that I don't like tools doing surprising or unintuitive things for inscrutable reasons. Balancing against that, however, are a couple points:

1. I _really_ dislike when tools give an error message that makes it clear they know what you intended, instead of doing the thing. We already have that a little when running `volta pin` outside of a package, but unifying the verbs but then having errors on `volta use npm@latest` that say `Did you mean "volta use --default npm@latest"` feels like going in the wrong direction.
2. We do have contextual behavior in the app already: The shims themselves! The core behavior of Volta is to provide this kind of contextual execution based on where you are when running a command.

That said, the point about `volta use node` is a great callout, since the "pin" case of that is the one entry point to _writing_ the `"volta"` key into an otherwise un-volta-fied `package.json` in the first place. The possibility of writing a config to a file _somewhere_, but it not being immediately obvious where, definitely feels like it will lead to hard-to-debug issues.

Also thinking out loud: I wonder if it would be clearer if we migrated to using a `.voltarc` for our "in-package" state and then made `volta use` only _create_ the file with a flag? That way the impact of calling `volta use` would be limited to modifying the existing `.voltarc` files, which very clearly indicate a connection with Volta.

Reply:
Yeah, I also share (1) though I also have come around on that over time because it's really easy for a tool to do the *wrong* thing on the assumption that the interpretation there is the correct one, and saying "You *might* have meant this; if so here's how to do it" increasingly seems better to me over time.

On (2) I thought I wrote as much above, but I didn't! I must have thought of it *later*. Shower mode brain or something. 😂 I think the difference, to me, is in the weird and admittedly extremely fuzzy distinction between “I am proactively trying to do something with Volta” in the `volta use` (or `pin` or whatever) case vs. “I am passively letting Volta act transparently on my behalf while doing something *else*” in the `node some-script.js` case. That is, again, *exceedingly* fuzzy, but I think it is part of what is motivating my intuition here. Take it for what it is worth. 😝

Reply:
Ultimately I agree that the contextual behavior has too many corner cases & opportunities for really poor UX. Updated the proposal to always use `volta pin` but with an explicit flag to indicate pinning the default version.